File: CS/RangeDate/MyGridControl/MyGridColumn.cs
// Developer Express Code Central Example:
// How to implement excel style tree filtering in a filter popup.
// 
// This example demonstrates how to filter dates using a tree and select date
// ranges.
// 
// You can find sample updates and versions for different programming languages here:
// http://www.devexpress.com/example=T143068

using System;
using System.Linq;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Columns;
using System.Collections.Generic;
using DevExpress.XtraGrid.Views.Base;

namespace DateRange {
    [System.ComponentModel.DesignerCategory("")]
    public class MyGridColumn : GridColumn {
        public new MyOptionsColumnFilter OptionsFilter { get { return base.OptionsFilter as MyOptionsColumnFilter; } }
        protected override OptionsColumnFilter CreateOptionsFilter() { return new MyOptionsColumnFilter(this); }
        protected override FilterPopupMode GetFilterPopupMode() {
            FilterPopupModeExtended modeExtended = OptionsFilter.FilterPopupMode;
            FilterPopupMode mode = FilterPopupMode.List;
            if(modeExtended == FilterPopupModeExtended.Default
                && (ColumnType.Equals(typeof(DateTime)) || ColumnType.Equals(typeof(DateTime?)))
                && FilterMode != ColumnFilterMode.DisplayText) modeExtended = FilterPopupModeExtended.DateSmart;
            if(modeExtended == FilterPopupModeExtended.Default) modeExtended = FilterPopupModeExtended.List;
            OptionsFilter.UseFilterPopupRangeDateMode = false;
            switch(modeExtended.ToString()) {
                case "Default": { mode = FilterPopupMode.Default; break; }
                case "List": { mode = FilterPopupMode.List; break; }
                case "CheckedList": { mode = FilterPopupMode.CheckedList; break; }
                case "Date": { mode = FilterPopupMode.Date; break; }
                case "DateSmart": { mode = FilterPopupMode.DateSmart; break; }
                case "DateAlt": { mode = FilterPopupMode.DateAlt; break; }
                case "DateRange": { mode = FilterPopupMode.Date; OptionsFilter.UseFilterPopupRangeDateMode = true; break; }
                default:
                    break;
            }
            return mode;
        }
        public bool PIsDateFilterPopup
        {
            get { return this.IsDateFilterPopup; }
        }
    }
    

    public enum FilterPopupModeExtended { Default, List, CheckedList, Date, DateSmart, DateAlt, DateRange }

    public class MyGridColumnCollection : GridColumnCollection {
        public MyGridColumnCollection(ColumnView view) : base(view) { }
        protected override GridColumn CreateColumn() {
            return new MyGridColumn();
        }
        public new MyGridColumn this[string fieldName] { get { return ColumnByFieldName(fieldName) as MyGridColumn; } }
        public new MyGridColumn this[int index] { get { return (MyGridColumn)List[index]; } }
    }
}

File: VB/RangeDate/MyGridControl/MyGridColumn.vb
' Developer Express Code Central Example:
' How to implement excel style tree filtering in a filter popup.
' 
' This example demonstrates how to filter dates using a tree and select date
' ranges.
' 
' You can find sample updates and versions for different programming languages here:
' http://www.devexpress.com/example=T143068

Imports System
Imports System.Linq
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Columns
Imports System.Collections.Generic
Imports DevExpress.XtraGrid.Views.Base

Namespace DateRange
	<System.ComponentModel.DesignerCategory("")>
	Public Class MyGridColumn
		Inherits GridColumn

		Public Shadows ReadOnly Property OptionsFilter() As MyOptionsColumnFilter
			Get
				Return TryCast(MyBase.OptionsFilter, MyOptionsColumnFilter)
			End Get
		End Property
		Protected Overrides Function CreateOptionsFilter() As OptionsColumnFilter
            Return New MyOptionsColumnFilter(Me)
		End Function
		Protected Overrides Function GetFilterPopupMode() As FilterPopupMode
			Dim modeExtended As FilterPopupModeExtended = OptionsFilter.FilterPopupMode
			Dim mode As FilterPopupMode = FilterPopupMode.List
			If modeExtended = FilterPopupModeExtended.Default AndAlso (ColumnType.Equals(GetType(Date)) OrElse ColumnType.Equals(GetType(Date?))) AndAlso FilterMode <> ColumnFilterMode.DisplayText Then
				modeExtended = FilterPopupModeExtended.DateSmart
			End If
			If modeExtended = FilterPopupModeExtended.Default Then
				modeExtended = FilterPopupModeExtended.List
			End If
			OptionsFilter.UseFilterPopupRangeDateMode = False
			Select Case modeExtended.ToString()
				Case "Default"
					mode = FilterPopupMode.Default
					Exit Select
				Case "List"
					mode = FilterPopupMode.List
					Exit Select
				Case "CheckedList"
					mode = FilterPopupMode.CheckedList
					Exit Select
				Case "Date"
					mode = FilterPopupMode.Date
					Exit Select
				Case "DateSmart"
					mode = FilterPopupMode.DateSmart
					Exit Select
				Case "DateAlt"
					mode = FilterPopupMode.DateAlt
					Exit Select
				Case "DateRange"
					mode = FilterPopupMode.Date
					OptionsFilter.UseFilterPopupRangeDateMode = True
					Exit Select
				Case Else
			End Select
			Return mode
		End Function
		Public ReadOnly Property PIsDateFilterPopup() As Boolean
			Get
				Return Me.IsDateFilterPopup
			End Get
		End Property
	End Class


	Public Enum FilterPopupModeExtended
		[Default]
		List
		CheckedList
		[Date]
		DateSmart
		DateAlt
		DateRange
	End Enum

	Public Class MyGridColumnCollection
		Inherits GridColumnCollection

		Public Sub New(ByVal view As ColumnView)
			MyBase.New(view)
		End Sub
		Protected Overrides Function CreateColumn() As GridColumn
			Return New MyGridColumn()
		End Function
		Default Public Shadows ReadOnly Property Item(ByVal fieldName As String) As MyGridColumn
			Get
				Return TryCast(ColumnByFieldName(fieldName), MyGridColumn)
			End Get
		End Property
		Default Public Shadows ReadOnly Property Item(ByVal index As Integer) As MyGridColumn
			Get
				Return CType(List(index), MyGridColumn)
			End Get
		End Property
	End Class
End Namespace
